File: backend/app/schemas/portfolio.py
from datetime import datetime

from pydantic import BaseModel, Field

from app.schemas.grid import ExplanationSchema


class PortfolioGridItem(BaseModel):
    numbers: list[int]
    stars: list[int] | None
    score: float


class PortfolioResponse(BaseModel):
    id: int
    game_id: int
    name: str
    strategy: str
    grid_count: int
    grids: list[PortfolioGridItem]
    diversity_score: float
    coverage_score: float
    avg_grid_score: float
    min_hamming_distance: float | None
    computed_at: datetime

    model_config = {"from_attributes": True}


class PortfolioGenerateRequest(BaseModel):
    grid_count: int = Field(10, ge=2, le=200)
    strategy: str = Field(
        "balanced",
        pattern=r"^(max_diversity|max_coverage|balanced|min_correlation)$",
    )
    profile: str = Field(
        "equilibre",
        pattern=r"^(equilibre|frequence|gaps|cooccurrence|balance)$",
    )
    weights: dict[str, float] | None = None


class PortfolioGenerateResponse(BaseModel):
    strategy: str
    grid_count: int
    grids: list[PortfolioGridItem]
    diversity_score: float
    coverage_score: float
    avg_grid_score: float
    min_hamming_distance: float | None
    computation_time_ms: float
    method_used: str
    explanation: ExplanationSchema | None = None

File: frontend/src/types/portfolio.ts
export interface PortfolioGridItem {
  numbers: number[];
  stars: number[] | null;
  score: number;
}

export interface PortfolioResponse {
  id: number;
  game_id: number;
  name: string;
  strategy: string;
  grid_count: number;
  grids: PortfolioGridItem[];
  diversity_score: number;
  coverage_score: number;
  avg_grid_score: number;
  min_hamming_distance: number | null;
  computed_at: string;
}

export interface PortfolioGenerateRequest {
  grid_count?: number;
  strategy?: string;
}

export interface PortfolioGenerateResponse {
  strategy: string;
  grid_count: number;
  grids: PortfolioGridItem[];
  diversity_score: number;
  coverage_score: number;
  avg_grid_score: number;
  min_hamming_distance: number | null;
  computation_time_ms: number;
  method_used: string;
}
